File: packages/engine/src/cpus/6502/instructions/rts.test.js
import { test, expect } from 'vitest'
import { BufferMapper } from '../../../buffer_mapper'
import { CPU } from '..'

test('rts', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  cpu.sp = 0xff
  mmu.data = [0x20, 0xac, 0x4] // JSR $04ac
  const g = cpu.run()
  for (let i = 0; i < 6; i++) g.next()
  expect(cpu.pc).toBe(0x4ac)
  expect(mmu.data[0x1ff]).toBe(0)
  expect(mmu.data[0x1fe]).toBe(2)

  mmu.data[0x4ac] = 0x60 // RTS
  for (let i = 0; i < 6; i++) g.next()
  expect(cpu.pc).toBe(0x3)
})

File: packages/engine/src/cores/famibob/cpu/instructions/rts.test.zig
const testing = @import("../testing/setup.zig");

test "RTS" {
    var cpu = testing.createCPU();
    cpu.sp = 0xff;
    cpu.write(0, 0x20);
    cpu.write(1, 0xac);
    cpu.write(2, 4);
    for (0..6) |_| {
        cpu.process();
    }
    try testing.expectEqual(0x4ac, cpu.pc);
    try testing.expectEqual(0, cpu.read(0x1ff));
    try testing.expectEqual(2, cpu.read(0x1fe));

    cpu.write(0x4ac, 0x60);
    for (0..6) |_| {
        cpu.process();
    }
    try testing.expectEqual(3, cpu.pc);
}
